<a href="https://colab.research.google.com/github/AnaliaLeyez/AnaliaLeyez/blob/main/TP2_Leyez_Anal%C3%ADa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TRABAJO PRÁCTICO 2
---  
El trabajo práctico consiste en responder una serie de preguntas utilizando mayormente las funciones de agrupamiento de pandas. Se utilizan dos archivos csv en google drive con información de emisión de gases de efecto invernadero provenientes de la FAO. Un archivo con datos correspondientes a todos los paises del mundo y el otro archivo solo con paises de América del Sur.

Los datos con los que se trabajan correponden al año 2020 y son una porcion del dataset "Emissions Totals" de FAOSTAT que resume las emisiones de gases de efecto invernadero (GHG) generadas por los sistemas agroalimentarios y que se difunden en los dominios de Emisiones del Cambio Climático de FAOSTAT.

Nota: Se recomienda revisar la descripción detallada del dataset en: ["Emissions Totals"](https://www.fao.org/faostat/es/#data/GT)

## INFORMACIÓN PRELIMINAR

La siguiente celda prepara dos dataframe **ghg_mundo** y **ghg_amesur** leyendo archivos ubicados en google drive que contienen una porción del dataset Totales de Emisión de la FAOSTAT.

**DataFrames**

* **ghg_mundo**: contienen los datos de emision de todos los paises del mundo
* **ghg_amesur**: contienen los datos de emision de los paises de america del sur

**Columnas**

* **Ambito** Dominio/ sección del dataset FAOSTAT, en este caso "Emisiones Totales"
* **Area**   Pais de origen de las emisiones
* **Elemento** Nombre del gas de efecto invernadero (GHG)
* **Producto** Nombre del item/producto/accion que genera la emisión
* **Año**      Año del valor informado
* **"Unidad** Unidad del valor informado (en miles de toneladas)
* **Valor** Valor informado

**Fuentes**

* **Origen** [FAOSTAT Emissions Totals"](https://www.fao.org/faostat/es/#data/GT)



# IMPORTS

In [ ]:
import pandas as pd
import numpy as np

# opciones de pandas
pd.options.display.max_columns = 20
pd.options.display.max_rows = 120
pd.options.display.min_rows = 80
pd.options.display.max_colwidth = 80

# DATA

In [ ]:
# Esta celda prepara los dataframes necesarios para responder los ejercicios del trabajo práctico.
url_mundo = "https://drive.google.com/u/1/uc?id=1n_pkLS9Tn0K6yclNTzwYlAWRdHRqZM3a&export=download"
url_amesur = "https://drive.google.com/u/1/uc?id=1vpEwDzzeiIiWM6-5LksiZNuKgy3DuDea&export=download"

# Preparamos los dos dataframe: ghg_mundo y ghg_amesur
ghg_mundo = pd.read_csv(url_mundo)
ghg_amesur = pd.read_csv(url_amesur)

# CONSIGNA Y RESOLUCIÓN

Utilizando los dataframe preparados arriba:

* **ghg_mundo**: con los datos a nivel del mundo
* **ghg_amesur**: con los datos de america del sur

Responder las siguientes preguntas y ejercicios:

## 1. Mostrar información general sobre las columnas, tipos de datos y cantidad de registros

In [ ]:
# 1. Mostrar información general sobre las columnas, tipos de datos y cantidad de registros

#Información general del DataFrame mundial
print("-------Información general del DataFrame ghg_mundo-------\n")
ghg_mundo.info()


#Información general del DataFrame de América del Sur
print("\n\n-------Información general del DataFrame ghg_amesur------\n")
ghg_amesur.info()

-------Información general del DataFrame ghg_mundo-------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11436 entries, 0 to 11435
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Ambito    11436 non-null  object 
 1   Area      11436 non-null  object 
 2   Elemento  11436 non-null  object 
 3   Producto  11436 non-null  object 
 4   Año       11436 non-null  int64  
 5   Unidad    11436 non-null  object 
 6   Valor     11436 non-null  float64
dtypes: float64(1), int64(1), object(5)
memory usage: 625.5+ KB


-------Información general del DataFrame ghg_amesur------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 710 entries, 0 to 709
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Ambito    710 non-null    object 
 1   Area      710 non-null    object 
 2   Elemento  710 non-null    object 
 3   Producto  710 non-null    object 
 4   Año       710 

__Conclusiones luego de correr el código.__

* <u>DataFrame ghg_mundo</u>

___Cantidad de registros:___ 11436

___Columnas:___ 7

___Nulos:___ No hay ya que la cantidad de non-null en cada columna coincide con el total de registros, por lo que no es necesario ver si eliminamos o cómo rellenamos los nulos.

Solo las columnas Año y Valor son de tipo numérico, las demás columnas son tipo object

---

* <u>DataFrame ghg_amesur</u>

___Cantidad de registros:___ 710

___Columnas:___ 7

___Nulos:___ No hay ya que la cantidad de non-null en cada columna coincide con el total de registros, por lo que no es necesario ver si eliminamos o cómo rellenamos los nulos.

Solo las columnas Año y Valor son de tipo numérico, las demás columnas son tipo object

## 2. Todos los valores informados tienen la misma unidad?

In [ ]:
# 2. Todos los valores informados tienen la misma unidad?

#Traigo los valores de la columna "Unidad" sin repetir, por lo que uso unique()
unidades_mundo = ghg_mundo['Unidad'].unique()
unidades_amesur = ghg_amesur['Unidad'].unique()

print("Las unidades de medición en ghg_mundo son: \n", unidades_mundo)
print("\n\nLas unidades de medición en ghg_amesur son: \n", unidades_amesur)


Las unidades de medición en ghg_mundo son: 
 ['mil toneladas']


Las unidades de medición en ghg_amesur son: 
 ['mil toneladas']


__Respuesta__

Sí, todos los valores informados tienen la misma unidad, tanto en ghg_mundo como en ghg_amesur. Todos se miden en "mil toneladas" (ya que al aplicar unique() no encontramos otra unidad), lo que facilita la comparación de mediciones.


## 3. Cuales son los gases de efecto invernadero (GHG) informados?

In [ ]:
# 3. Cuales son los gases de efecto invernadero (GHG) informados?

#Traigo los valores de la columna "Elemento" sin repetir, por lo que uso unique()
gases_mundo = ghg_mundo['Elemento'].unique()
gases_amesur = ghg_amesur['Elemento'].unique()

print("Los gases de efecto invernadero (GHG) informados en ghg_mundo son: \n", gases_mundo)
print("\n\nLos gases de efecto invernadero (GHG) informados en ghg_mundo son: \n", gases_amesur)

Los gases de efecto invernadero (GHG) informados en ghg_mundo son: 
 ['Emisiones (N2O)' 'Emisiones (CH4)' 'Emisiones (CO2)']


Los gases de efecto invernadero (GHG) informados en ghg_mundo son: 
 ['Emisiones (N2O)' 'Emisiones (CH4)' 'Emisiones (CO2)']


__Respuesta__
Se consultó el DataFrame con los datos de América del Sur y del mundo para entender mejor si hubiera gases que se informan a nivel mundial pero no en América del Sur.
Se observa que en ambos casos los gases informados son los mismos, Estos son:

N2O, CH4 y CO2

## 4. Cuales son los "Productos" que cuentan en la emisión de GHG a nivel mundial ?

In [ ]:
# 4. Cuales son los "Productos" que cuentan en la emisión de GHG a nivel mundial ?

#Traigo los valores de la columna "Producto" sin repetir, por lo que uso unique()
productos_mundo = ghg_mundo['Producto'].unique()

print("Los productos que cuentan en la emisión de GHG a nivel mundial son: \n", productos_mundo)

Los productos que cuentan en la emisión de GHG a nivel mundial son: 
 ['Residuos agrícolas' 'Cultivo del arroz' 'Quemado de residuos agrícolas'
 'Fermentación entérica' 'Gestión del estiércol'
 'Estiércol depositado en las pasturas' 'Estiércol aplicado a los suelos'
 'Fertilizantes sintéticos' 'Suelos orgánicos drenados (CO2)'
 'Suelos orgánicos drenados (N2O)' 'Tierras forestales'
 'Conversión neta de bosques' 'Incendios de sabana'
 'Incendios en suelos de turba' 'Incendios forestales'
 'Incendios en los bosques tropicales húmedos'
 'Eliminación de desechos de sistemas agroalimentarios'
 'Fabricación de fertilizantes' 'Uso de electricidad en agrícola'
 'Venta de alimentos' 'Consumo\xa0de alimentos en los hogares'
 'Transformación\xa0de alimentos' 'Energía' 'IPPU' 'Desechos' 'Otro'
 'On-farm energy use' 'Fabricación de pesticidas' 'Envasado alimentario'
 'On-farm Heat Use']


## 5. Cual es el total de emision mundial ?

In [ ]:
# 5. Cual es el total de emision mundial ?

#Sumo todos los valores de la columna "Valor"
total_mundial = ghg_mundo['Valor'].sum().round(2)

print(f"El total de emisión mundial es: {total_mundial} mil toneladas")

El total de emisión mundial es: 52464596.38 mil toneladas


## 6. Cuales son los paises que más emiten y cuales los que menos emiten (mostrar 3 por cada caso)?

In [ ]:
# 6. Cuales son los paises que más emiten y cuales los que menos emiten (mostrar 3 por cada caso)?

#Sumo el valor emitido, habiendo agrupado por área (país)
emision_total_por_pais = ghg_mundo.groupby('Area')['Valor'].sum().round(2).reset_index()

#Países que más emiten, ordeno de menor a mayor y tomo los últimos 3
paises_mayor_emision = emision_total_por_pais.sort_values(by='Valor').tail(3)

#Países que menos emiten, ordeno de menor a mayor y tomo los primeros 3
paises_menor_emision= emision_total_por_pais.sort_values(by='Valor').head(3)

print("Países que más emiten:\n", paises_mayor_emision)
print("\n\nPaíses que menos emiten:\n", paises_menor_emision)

Países que más emiten:
                          Area        Valor
64  Estados Unidos de América   4752453.97
41         China, Continental  11822996.65
40                      China  12167542.99


Países que menos emiten:
             Area    Valor
76         Ghana -1214.17
174  Puerto Rico  -738.63
82          Guam  -158.18


__Observación:__

Debido a que los paises que menos emiten registran un valor negativo, busqué si esto era posible o debía elliminar dichos registros. Encontré que efectivamente es posible y sería un indicativo de queel paí está removiendo más gases de efecto invernadero de los que emite.

## 7. Cuál es la cantidad total de cada gas emitidos a nivel mundial?

In [ ]:
# 7. Cuales es la cantidad total de cada gas emitidos a nivel mundial?

#Sumo el valor emitido, habiendo agrupado por Elemento
emision_mundial_por_elemento = ghg_mundo.groupby('Elemento')['Valor'].sum().round(2).reset_index()

print("Cantidad total de cada gas emitido a nivel mundial: \n", emision_mundial_por_elemento)

Cantidad total de cada gas emitido a nivel mundial: 
           Elemento        Valor
0  Emisiones (CH4)    479800.69
1  Emisiones (CO2)  51971063.19
2  Emisiones (N2O)     13732.50


## 8. En américa del sur, cuales son los valores totales de cada gas emitido?

In [ ]:
# 8. En américa del sur, cuales son los valores totales de cada gas emitido?

#Sumo el valor emitido, habiendo agrupado por Elemento
emision_amesur_por_elemento = ghg_amesur.groupby('Elemento')['Valor'].sum().round(2).reset_index()

print("Cantidad total de cada gas emitido en América del Sur: \n", emision_amesur_por_elemento)

Cantidad total de cada gas emitido en América del Sur: 
           Elemento       Valor
0  Emisiones (CH4)    43876.10
1  Emisiones (CO2)  1941783.70
2  Emisiones (N2O)     1293.21


América del Sur representa al rededor del 10% de las emisiones mundiales de gases GHG.

## 9. Basado en 7 y 8, indicar que porcentaje de cada gas corresponde a américa latina respecto del mundo.

In [ ]:
# 9. Basado en 7 y 8, indicar que porcentaje de cada gas corresponde a américa latina respecto del mundo.

#Calculo el porcentaje usando las variables ya declaradas en los puntos 7 y 8
porcentaje = (emision_amesur_por_elemento['Valor'] / emision_mundial_por_elemento['Valor']) * 100
porcentaje = porcentaje.round(2)

#Creo una copia del Df usado en el punto 8 para luego sumarle el valor mundial y la nueva columna Porcentaje
amesur_por_elemento = emision_amesur_por_elemento.copy()

#Renombro 'Valor' por 'Valor AmeSur' para mayor claridad y distinción
amesur_por_elemento.rename(columns={'Valor': 'Valor AmeSur'}, inplace=True)

#Sumo la columna que muestra el valor mundial, para ver mejor la comparativa
amesur_por_elemento['Valor mundial'] = emision_mundial_por_elemento['Valor']

#Sumo una columna de porcentaje al Df
amesur_por_elemento['Porcentaje AmeSur/Mundial'] = porcentaje

print("Porcentaje de emisión de cada gas en América del Sur respecto al mundo:\n", amesur_por_elemento)

Porcentaje de emisión de cada gas en América del Sur respecto al mundo:
           Elemento  Valor AmeSur  Valor mundial  Porcentaje AmeSur/Mundial
0  Emisiones (CH4)      43876.10      479800.69                       9.14
1  Emisiones (CO2)    1941783.70    51971063.19                       3.74
2  Emisiones (N2O)       1293.21       13732.50                       9.42


## 10. En america del sur, que productos emiten más GHG y cuales menos (mostrar todos)?

In [ ]:
# 10. En america del sur, que productos emiten más GHG y cuales menos (mostrar todos)?

#Sumo las emisiones habiendo agrupado por Producto
emision_amesur_por_producto = ghg_amesur.groupby('Producto')['Valor'].sum().round(2).reset_index()

#Ordeno de mayor a menor emisión
productos_amesur_ordenados = emision_amesur_por_producto.sort_values(by='Valor', ascending=False)

print("Emisión de GHG por producto en América del Sur:\n", productos_amesur_ordenados)

Emisión de GHG por producto en América del Sur:
                                                 Producto       Valor
1                             Conversión neta de bosques  1085941.54
5                                                Energía   931390.30
14                                                  IPPU    77449.39
26                           Transformación de alimentos    65944.70
19                                    On-farm energy use    41053.28
0                    Consumo de alimentos en los hogares    40681.89
11                                 Fermentación entérica    22081.44
10                             Fabricación de pesticidas    17907.00
23                       Suelos orgánicos drenados (CO2)    13230.91
6                                   Envasado alimentario    10550.78
28                                    Venta de alimentos     9726.59
9                           Fabricación de fertilizantes     9469.98
3                                               Desech

## 11. Y en el mundo..., que productos emiten más GHG y cuales menos (mostrar todos)?

In [ ]:
# 11. Y en el mundo..., que productos emiten más GHG y cuales menos (mostrar todos)?

#Sumo las emisiones habiendo agrupado por Producto
emision_mundo_por_producto = ghg_mundo.groupby('Producto')['Valor'].sum().round(2).reset_index()

#Ordeno de mayor a menor emisión
productos_mundo_ordenados = emision_mundo_por_producto.sort_values(by='Valor', ascending=False)

print("Emisión de GHG por producto en el mundo:\n", productos_mundo_ordenados)

Emisión de GHG por producto en el mundo:
                                                 Producto        Valor
5                                                Energía  41762246.97
14                                                  IPPU   4464238.14
1                             Conversión neta de bosques   2945052.57
0                    Consumo de alimentos en los hogares   1634384.62
27                           Transformación de alimentos   1247063.11
24                       Suelos orgánicos drenados (CO2)    833598.71
20                                    On-farm energy use    518846.73
9                           Fabricación de fertilizantes    492550.80
28                       Uso de electricidad en agrícola    485421.93
6                                   Envasado alimentario    465566.46
29                                    Venta de alimentos    319850.74
3                                               Desechos    122410.69
11                                 Fermentación 

## 12. En América del Sur, cuales son los paises que más y menos emiten (mostrar 3 por cada caso)?

In [ ]:
# 12. En América del Sur, cuales son los paises que más y menos emiten (mostrar 3 por cada caso)?

#Sumo el valor emitido, habiendo agrupado por área país, usando del Df de América del Sur
emision_amesur_por_pais = ghg_amesur.groupby('Area')['Valor'].sum().round(2).reset_index()

#Países que más emiten, ordeno de menor a mayor y tomo los últimos 3
paises_mayor_emision_amesur = emision_amesur_por_pais.sort_values(by='Valor').tail(3)

#Países que menos emiten, ordeno de menor a mayor y tomo los primeros 3
paises_menor_emision_amesur= emision_amesur_por_pais.sort_values(by='Valor').head(3)

print("Países que más emiten:\n", paises_mayor_emision_amesur)
print("\n\nPaíses que menos emiten:\n", paises_menor_emision_amesur)

Países que más emiten:
         Area      Valor
4   Colombia  195829.08
0  Argentina  253346.17
2     Brasil  958774.81


Países que menos emiten:
                 Area     Valor
6   Guayana Francesa   2782.01
11           Uruguay   9080.53
10          Suriname  12468.77
